### ISSUE TITLE
[Feature Request] Screen reader accessibility

### ISSUE THREAD

**Issue by AJenbo** [libsdl-org/SDL#7463]
I am currently working on [making DevilutionX accessible for blind players](https://github.com/diasurgical/devilutionX/pull/5866) in consultation with a blind person who has been involved in making several other games accessible. One big stumbling block is interacting with screen readers. SDL already provides easy access to gamepads, a stereo sound system, sensors and vibrators, so adding screen reader support doesn't seem out of place and is the missing piece in the puzzle for making games accessible to the visually impaired.

Several people from the blind community have suggested that I use a library called [Tolk](https://github.com/sig-a11y/tolk/pull/2). While it does provide a simple interface for a range of different screen readers, it has some major downsides. It only supports Windows, it expects the use of MSVC, and the project appears to be abandoned.

I have also found another similar library that seems a little better but also has many of the same shortcomings. It's API is interesting in that it provides a lot more control over the screen readers: https://github.com/qtnc/UniversalSpeech

Rather than forking one of these and porting it, or creating yet another project, I think the community would be much better served by having this feature integrated directly into SDL, as it could remove the barrier for adding the needed aids, so I hope you find this relevant for the project.

For a minimal API, something like this would be needed:
```cpp
void SDL_ReaderSpeak(const char*, bool interrupt);
bool SDL_ReaderIsSpeaking();
bool SDL_ReaderIsEnabled();
void SDL_ReaderStop();
``` 

P.S. I am willing to work on this given a bit of guidance here and there on what the API should look like and how it would preferably be done. I have already managed to get screen reader integration working on Linux. And between the two mentioned libraries I think I have some good examples of how to implement this for Windows as well. Apple also has its own API, and I expect this is also the case for other platforms like Android, Xbox and PlayStation.

P.S.S. Screen Readers are also helpful for dyslexic players

**Comment by miss-programgamer** [comment]
This is a nice initiative, but I think the approach of making a separate, well-maintained library for interacting with screen readers is the way to go, rather than integrating it into SDL directly. If I was a developer looking for such an API (which I'll probably be in the future), I wouldn't want to be forced to include the entirety of SDL just to access that functionality if that's all I wanted to use. I think making a standalone library, then adding an extension lib to SDL in the vein of SDL_ttf and SDL_image (something like SDL_reader) would serve everyone better in the long run.

**Comment by AJenbo** [comment]
SDL_reader is a good suggestion, though I don't have the time to manage either of such projects.

**Comment by inkydragon** [comment]
The [AccessKit/accesskit](https://github.com/AccessKit/accesskit) project has an ambitious goal: UI accessibility infrastructure across platforms and programming languages.

`accesskit` is written in rust and is in very early stages of development currently (2023-03).
There is only a working POC on Windows, other platforms are still under development.

**Comment by slouken** [comment]
https://github.com/libsdl-org/SDL/issues/4225 is related

**Comment by slouken** [comment]
I'd be interested to see an API proposal. I'll go ahead and close this, but feel free to reopen it if you'd like to move forward with this.

**Comment by loftafi** [comment]
How would one know when to call `SDL_ReaderSpeak`?  First we would need to know if the OS (Windows, iOS, etc...) has accessibility mode enabled. It would also be necessary to know what accessibility features the user has requested, i.e.

```
mode = SDL_GetAccessibilityMode()
if (mode == LARGE_TEXT) { ... }
if (mode == EXTRA_LARGE_TEXT) { ... }
if (mode == TEXT_TO_SPEECH) { ... }
```

**Comment by AJenbo** [comment]
Hm for what I had I would just call it and the OS would then ignore it if text to speech was not enabled by the user.

**Comment by loftafi** [comment]
@AJenbo  That makes sense.  The use case I am thinking of, are the people I know who have some very very very limited eyesight. They actually use a combination of large text and screen readers to navigate their phone.  So I guess I am thinking of accessibility in a broader sense of not just blind, but limited eyesight, deaf, and or limited movement (cant use a keyboard with their hands)

The second thing is, if you want users to be able to have labels read. Then you need the users to be able to "tab" through labels. So it is necessary for the game to know if accessibility is "on" so that it knows if "tab" (or whatever navigation device is being used) is intended to go to the next "clickable element" or if "tab" should go to the next "readable or clickable element"

The way this works in Godot (and I assume other engines) is that you can set an accessibility status on a label to allow a label to be "tabbed into."

**Comment by AJenbo** [comment]
That's make sense, my thought was that the screen Reader would allow them to navigate to settings where they could enable further options

**Comment by loftafi** [comment]
I guess it depends on the target platform. I was mostly thinking about phones, where TTS, UI, color settings are phone wide settings. I guess consoles might not have the same type of menus, Im not a console person myself.

**Comment by AJenbo** [comment]
I'm mostly focused on desktop 🙂

**Comment by rubenlg** [comment]
Having SDL abstract away some of the OS a11y features would be very convenient for making our games more accessible. The web provides a standardized set of semantics that offer cross-platform accessibility support and can serve as a strong source of inspiration.

I'm not an accessibility expert, but I've used a11y's features while developing several websites, so I'll share what I know:

The most important component is the accessibility tree, which the web platform does not expose directly to developers. Instead, the browser builds it from the DOM, with developers using ARIA attributes to indicate which nodes should participate in the a11y tree and how. Given the many different ways UI can be represented and navigated in SDL-based games and applications, the web’s approach is not directly applicable. A more suitable API for SDL would likely allow creating and updating an explicit accessibility tree manually.

For focus management (e.g. tabbing), bi-directional communication is key: the game should be able to notify the screen reader when a different node gains focus, and the screen reader should be able to request focus changes as well.

The web also supports user a11y preferences such as high contrast mode, reduced motion, and color schemes through CSS media queries. In a native context like SDL, these could be mapped to simple getter functions.

**Comment by AJenbo** [comment]
> A more suitable API for SDL would likely allow creating and updating an explicit accessibility tree manually.

That's a really good suggestion 👍

**Comment by loftafi** [comment]
Games typically already have a list of objects that a user can tab through. I don't think there is a need for a specialized "accessibility tree." My suggestion is this is the bare bones minimum that would be needed as a basic starting point:

- the ability to detect if the user (via the OS) has requested accessibility to tab into labels,
- the ability to detect if the user (via the OS) has requested vision support (large font size)
- the ability to trigger the OS to read text content (the text of a label, the name of a character, etc...)

Am I missing something?

    void SDL_ScreenReaderSpeak(const char*, bool interrupt);
    bool SDL_ScreenReaderStatus();  //enabled or disabled 
    int SDL_OSTextSize();  // normal, larger 

Events? I don't know if this is possible, I don't know if anyone would need it, but maybe???

    SDL_ScreenReader_Enabled
    SDL_ScreenReader_Disabled
    SDL_ScreenReader_Speaking 
    SDL_ScreenReader_FinishedSpeaking 

I am not sure if speaking and finished speaking are things that people would need and if the OS even shares this type of event? Im not sure if I would have any use for started/finished speaking.

**Comment by loftafi** [comment]
>  the screen reader should be able to request focus changes as well.

On this specifically, do you mean when a screen reader is connected to a phone or computer it can send "tab" events to the OS that SDL perceives as a "tab" keyboard event. Or are you saying there a special non-keyboard OS "tab" type event that screen readers can send that SDL does not already make accessible via the SDL api? 

I was under the impression (perhaps incorrectly) that screen readers for blind people, and mobility limited accessibility controllers would be sending standard OS events (tab, left, right, enter, etc...) Is that not correct?

**Comment by rubenlg** [comment]
> Games typically already have a list of objects that a user can tab through. I don't think there is a need for a specialized "accessibility tree." My suggestion is this is the bare bones minimum that would be needed as a basic starting point:

The accessibility tree is a standard component of assistive technologies. It already exists, it's not something SDL needs to invent. What SDL can do is provide a unified, cross-platform API for populating this tree, since each OS exposes it differently.

Note that the accessibility tree is separate from your game’s object structure. Even on the web, the accessibility tree differs from the DOM (though it's computed from it).

> void SDL_ScreenReaderSpeak(const char*, bool interrupt);

In general, it's not ideal to force screen readers to speak text at specific times (and it's not even clear that's reliably possible without hacks). A better approach is to declare your accessibility tree and let the user navigate it freely. When navigating the tree, a screen reader may re-announce nodes to help users understand their current context. For example, in a dialogue menu, users might want to explore all dialogue options in any order (and even repeat a single one or all of them) before making a choice. Some screen readers can also detect when the screen has changed significantly and begin reading from the top, though users can interrupt and navigate manually, triggering partial reads. Screen readers do a lot of smart things based on the tree, and users who rely on them depend on that flexibility and control.

If you're voicing characters using recorded actor dialogue or synthesized speech directly, you may not need accessibility integration for that. However, if you want screen readers to announce those lines, there are ways to inject temporary, high-priority elements into the accessibility tree. Even then, the screen reader might delay reading them until it finishes another sentence. And the user should still be able to jump to the settings menu midway through a spoken line (just as a mouse user would) if the UI allows it.

> I was under the impression (perhaps incorrectly) that screen readers for blind people, and mobility limited accessibility controllers would be sending standard OS events (tab, left, right, enter, etc...) Is that not correct?

Not quite. When a screen reader or other assistive tech connects to the OS using the standard a11y APIs, it operates at a higher semantic level. Instead of simulating low-level key events (like Tab or Left Arrow), it sends high-level actions tied to specific nodes in the accessibility tree, such as "Activate node X"," "Press node Y (if declared as a button)" or "Increment node Z (if declared as a slider)."

Even if it were possible to simply tell the screen reader to speak a line immediately, and you implemented navigation yourself, you'd essentially be using it as a text-to-speech engine, leaving yourself responsible for implementing your custom screen reader behavior, which is far more complex than it might appear on the surface. For example, ChromeVox (an open-source screen reader) is over 100,000 lines of code (of which the TTS is less than 5,000), and it's not even the most complex one.

**Comment by LukeDickerson19** [comment]
+1 for SDL_Reader, this is the reason I'm unfortunately returning to dart/flutter for cross platform app development.

Perhaps instead of a new library such as SDL_Reader, the accessibility tree could be a feature added onto a pre-existing UI component lib such as [imgui](https://github.com/ocornut/imgui), [nukklear](https://github.com/Immediate-Mode-UI/Nuklear), or [Clay](https://github.com/nicbarker/clay). I think the fields of each node in the accessibility tree (role, label, state, etc.) could be struct members in each UI component, however these libs don't handle cross platform accessibility integration, so maybe they're output could include an accessibility tree, and then SDL_Reader passes it to the underlying platform? idk ...

Mobile app store accessibility requirements are often stricter for mobile apps than games cause they can be used for banking, messaging, transport, government, etc.
